pkg/controller/dns/letsencrypt.go

```diff
@@ -0,0 +1,295 @@
+
+	if err := client.Challenge.SetDNS01Provider(dnsProvider); err != nil {
+		return nil, err
+	}
+
+	request := certificate.ObtainRequest{
+		Domains: []string{fmt.Sprintf("*.%s", strings.TrimPrefix(domain, "."))},
+		Bundle:  true,
+	}
+
+	return client.Certificate.Obtain(request)
+}
+
+func EnsureLEUser(ctx context.Context, client kclient.Client, domain string) (*LEUser, error) {
+
+	targetUsername := strings.TrimPrefix(domain, ".") // leading dot is an issue especially for email addresses
+
+	leAccountSecret := &corev1.Secret{}
+	err := client.Get(ctx, router.Key(system.Namespace, system.LESecretName), leAccountSecret)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	// Existing LE User Secret found
+	if err == nil {
+		leUser, err := FromSecret(leAccountSecret)
+		if err != nil {
+			return nil, err
+		}
+
+		// Domain changed, recreate the secret
+		if !strings.Contains(leUser.Name, targetUsername) {
+			logrus.Infof("deleting LE secret for domain %s", domain)
+			if err := client.Delete(ctx, leAccountSecret); err != nil {
+				return nil, err
+			}
+		} else {
+			return leUser, nil
+		}
+	}
+
+	// Create and Register Let's Encrypt User
+	leUser := &LEUser{
+		Name:  targetUsername,
+		Email: fmt.Sprintf("%s@on-acorn.io", targetUsername),
+		URL:   LEEndpoint,
+	}
+	if err := leUser.Register(); err != nil {
+		return nil, fmt.Errorf("problem registering Let's Encrypt User: %w", err)
+	}
+
+	sec, err := leUser.ToSecret()
+	if err != nil {
+		return nil, fmt.Errorf("problem creating Let's Encrypt User secret: %w", err)
+	}
+
+	if err := client.Create(ctx, sec); err != nil {
+		return nil, fmt.Errorf("problem creating Let's Encrypt User secret: %w", err)
+	}
+
+	logrus.Infof("Registered Let's Encrypt User: %s", leUser.Name)
+
+	return leUser, nil
+
+}
+
+func (u *LEUser) EnsureWildcardCertificateSecret(ctx context.Context, client kclient.Client, dnsendpoint, domain, token string) (*corev1.Secret, error) {
+
+	sec := &corev1.Secret{}
+	err := client.Get(ctx, router.Key(system.Namespace, system.TLSSecretName), sec)
+	if err != nil && !apierrors.IsNotFound(err) {
+		// error fetching the existing secret, but it could exist
+		return nil, err
+	}
+
+	if err == nil {
+		if sec.Labels[labels.AcornDomain] != domain {
+			logrus.Infof("domain changed, renewing wildcard certificate")
+		} else {
+			// check if certificate is still valid
+			x509crt, err := certcrypto.ParsePEMCertificate([]byte(sec.Data[corev1.TLSCertKey]))
+			if err != nil {
+				logrus.Errorf("problem parsing existing TLS secret: %v", err)
+				// continues to generate new cert
+			} else {
+				if x509crt.NotAfter.After(time.Now().UTC()) {
+					// cert is still valid, return it
+					logrus.Infof("existing TLS secret %s is still valid until %s", x509crt.Subject.CommonName, x509crt.NotAfter)
+					return sec, nil
+				} else {
+					logrus.Infof("existing TLS secret %s is expired since %s", x509crt.Subject.CommonName, x509crt.NotAfter)
```

Comment:
Is this how renew works? If so you need to renew the cert before it expires. So something like a week before. Also you can't delete the secret until a new cert is obtained to avoid downtime.